sci-libs/gdal: Bump to v2.0.3

Whissi opened the pull request:
- Version bump to new upstream release v2.0.3.
- Bumped to EAPI=6.
- "oracle" USE flag added (Fixes Gentoo-Bug #[585156](https://bugs.gentoo.org/585156)).
- Fixes a build issue when using "java" USE flag and you have a JRE
  installed.
- Switched JPEG-2000 provider from media-libs/jasper to
  media-libs/openjpeg:2 because JasPer is unmaintained and vulnerable.

Whissi commented:
PR updated to replace weird for loop calling `newman` with a single `doman` call.

SoapGentoo commented on sci-libs/gdal/gdal-2.0.3.ebuild:
```diff
@@ -0,0 +1,313 @@
+WANT_AUTOCONF="2.5"
+
+GENTOO_DEPEND_ON_PERL="no"
+PYTHON_COMPAT=( python2_7 python3_{3,4} )
```
Please check py3.5 compatibility

Whissi commented:
Python3_5 supported enabled like suggested by @SoapGentoo. Haven't tried each provided Python script however all Python3_4 deps are also available for Python3_5 and no module was found listed in Python's 3.5 migration guide which require some action.

Whissi commented:
Rebased to fix QA check which was broken due to the xtrabackup-bin/libgcrypt update.

SoapGentoo commented on sci-libs/gdal/gdal-2.0.3.ebuild:
```diff
@@ -0,0 +1,313 @@
+	use perl && perl_delete_localpod
+
+	local DOCS=( Doxyfile HOWTO-RELEASE NEWS )
+	use doc && HTML_DOCS=( html/* )
```
prefer `use doc && HTML_DOCS=( html/. )` in order to prevent massive globbing

SoapGentoo commented on sci-libs/gdal/gdal-2.0.3.ebuild:
```diff
@@ -0,0 +1,313 @@
+		pushd "${S}"/swig/perl > /dev/null
+		perl-module_src_configure
+		perl-module_src_compile
+		popd > /dev/null
```
missing `|| die`

SoapGentoo commented on sci-libs/gdal/gdal-2.0.3.ebuild:
```diff
@@ -0,0 +1,313 @@
+	default
+
+	if use perl ; then
+		pushd "${S}"/swig/perl > /dev/null
```
missing `|| die`

SoapGentoo commented on sci-libs/gdal/gdal-2.0.3.ebuild:
```diff
@@ -0,0 +1,313 @@
+
+src_install() {
+	if use perl ; then
+		pushd "${S}"/swig/perl > /dev/null
```
missing `|| die`

SoapGentoo commented on sci-libs/gdal/gdal-2.0.3.ebuild:
```diff
@@ -0,0 +1,313 @@
+	if use perl ; then
+		pushd "${S}"/swig/perl > /dev/null
+		perl-module_src_install
+		popd > /dev/null
```
missing `|| die`

Whissi commented:
Thanks! PR updated. But the JVM problem is still unresolved.

SoapGentoo commented:
@Whissi remind me again, whats the JVM problem that you are referring to?

Whissi commented:
1. Build with "mdb" USE flag (require JAVA).
2. Now libgdal.so will be linked against libjvm.so
...but the system cannot resolve libjvm.so.

I already discussed the problem with @chewi in #gentoo-java: A system can have two different JVMs: A **system** and **user** JVM. So `eselect java-vm` can't create a symlink like `eselect postgresql` can do for postgresql's libpg.so.

The ebuild could enforce a JVM but this would remove the possibility to switch JVM at runtime.

A wrapper isn't an option because everything which uses *libgdal.so* would require the same wrapper.

Using `/etc/java-config-2/current-system-vm/` is also not an option because the path is different between JRE and JDK usage.

@chewi came up with the idea to patch gdal to use `dlopen()`. This way we could add a logic which would first check the path and then load the right *libjvm.so* but I don't know if he had time to look into this yet.

chewi commented:
I've had a go at this and it works. You can see the commit against master at chewi/gdal@5351b060d6d859e96d1a2f40c6b3e942cc755abd. It more or less applies against 2.0.3 but please note that 2.1.1 came out last month.

Unfortunately the experience still isn't perfect out of the box because the jackcess jar plus its dependencies are needed in the CLASSPATH. Setting this manually is less hazardous than setting LD_LIBRARY_PATH and it's less likely to change over time. We should get jackcess packaged though. To polish this off nicely, we could patch it a little further so that it sets the classpath automatically under Gentoo.

I've already spent longer on this than I really wanted to so I'd appreciate it if you could take it from here by testing it further and making the necessary adjustments. You'll need [jackcess 1.2.14.3](https://sourceforge.net/projects/jackcess/files/jackcess/1.2.14.3/jackcess-1.2.14.3.jar/download) and you'll need to emerge commons-logging, commons-lang:2.1, commons-codec, and log4j. You'll also need some [test data](http://www.state.nj.us/transportation/refdata/countysidewalks/zip/mdb/Somerset.zip). Then do this.

```
CLASSPATH=~/Downloads/jackcess-1.2.14.3.jar:$(java-config -d -p commons-logging,commons-lang-2.1,commons-codec,log4j) ogrinfo ~/Downloads/Somerset.mdb
```

chewi commented:
I made a change that allows it to work even if JAVA_HOME isn't set by falling back to java in the PATH. This is ideal for Gentoo as JAVA_HOME is usually only set by shell profiles. /usr/bin/java will work in any context. The commit reference above has been updated.

I've also pushed chewi/gdal@df5214dc698ad0e48c71b3739a437d4d9579a972 in a separate gentoo branch that fetches the jackcess classpath automatically using java-config. This assumes that jackcess 1.2.14.3 has been given a SLOT of 1. v2 is totally incompatible so I wouldn't bother packaging it at the moment.

Once we've closed this out and you're happy with it, I'll file a pull request upstream.

monsieurp commented:
what's the deal with this PR? can we move this forward?

chewi commented:
I too would like to close this one out as I did work hard on it.

Whissi commented:
I still have to pack some JAVA packages. Will look into this later this week.

mgorny commented:
Ping here.

Amynka commented:
As from I get @Whissi  wants to work on it more.. Feel free to commit it when it's done.

mgorny commented:
@Amynka, yeah, I wanted to ping him ;-).

Whissi commented:
It's been a while but here's the update.

Like @chewi said I created an ebuild for jackcess. Could anyone from @gentoo/java please review?

@chewi: I backported your gdal changes to v2.0.x branch. I had to fix https://github.com/Whissi/gdal/commit/00c8cc8b2a678c4104e802bfbaccb60fbb55e4d3#diff-964416868e77c518b6b5c8b2b059c803R222 due to https://travis-ci.org/Whissi/gdal/builds/181060588 -- please review.

Changes from https://github.com/gentoo/gentoo/pull/2423 should be incorporated.

This ebuild works for me and I am able to use JAVA-enabled applications like *ogrinfo*.

chewi commented on dev-java/jackcess/jackcess-1.2.14.3.ebuild:
```diff
@@ -0,0 +1,44 @@
+RDEPEND=">=virtual/jre-1.6
+	${CDEPEND}"
+
+JAVA_GENTOO_CLASSPATH="
```
You can remove `JAVA_GENTOO_CLASSPATH` entirely by simply renaming `CDEPEND` to `CP_DEPEND`. It's a new trick.

chewi commented on dev-java/jackcess/jackcess-1.2.14.3.ebuild:
```diff
@@ -0,0 +1,44 @@
+
+src_compile() {
+	java-pkg-simple_src_compile
+	java-pkg_addres "${PN}.jar" . \( \
```
The extra find arguments are not needed because the only other files contained in the source jar are automatically ignored by java-pkg_addres. It won't hurt if an extra file slips in later but it will hurt if you miss one.

chewi commented on dev-java/jackcess/jackcess-1.2.14.3.ebuild:
```diff
@@ -0,0 +1,44 @@
+	>=dev-java/commons-logging-1.1.3:0
+	>=dev-java/log4j-1.2.7:0"
+
+DEPEND=">=virtual/jdk-1.6
```
Always `DEPEND` on at least the oldest Java we have in the tree (1.7) and ideally test with this version as well as 1.8. Same goes for `virtual/jre` below.

Whissi replied:
Do we prefer ">=virtual/jdk-1.7:=", i.e. do we want an automatic rebuild when virtual/jdk changes or ">=virtual/jdk-1.7:*", i.e. no automatic rebuild?

chewi replied:
You don't normally need either because of an exception in repoman's rules though I am thinking of changing that. Definitely the latter, if you find you need the former than you're doing it wrong.

chewi commented on sci-libs/gdal/gdal-2.0.3.ebuild:
```diff
@@ -0,0 +1,294 @@
+	java? ( >=virtual/jre-1.6:* )
+	jpeg? ( virtual/jpeg:0= )
+	jpeg2k? ( media-libs/openjpeg:2= )
+	mdb? ( dev-java/jackcess )
```
Always add a SLOT when dealing with Java dependencies. This should be `dev-java/jackcess:1`.

chewi commented on sci-libs/gdal/gdal-2.0.3.ebuild:
```diff
@@ -0,0 +1,294 @@
+	spatialite? ( sqlite )
+"
+
+pkg_setup() {
```
You can just leave this out entirely?

Whissi replied:
No, looks like this is need. Otherwise *configure* will fail with
```
checking whether we should include Java support... configure: error: "Cannot find /opt/oracle-jre-bin-1.8.0.112/include or <mac_java_framework>/Headers directory."
```

chewi replied:
That's strange. It's exported by java-pkg-opt-2. I think there's a pkg_setup exported by the Python stuff (hard to tell) that you should probably be calling too. It won't be overriding the Java one though as java-pkg-opt-2 is last in the inherit list.

I tried it myself by doing `USE="java mdb" ebuild gdal-2.0.3.ebuild clean setup` and, assuming you have `virtual/jdk` rather than `virtual/jre`, you see the line `Using: icedtea-bin-8` whether `pkg_setup` is present or not. It only goes away if you have a blank `pkg_setup` so it's definitely being called. Something weird with your test?

Whissi replied:
No, you are right. When I tested I still had only *virtual/jre* set for *RDEPEND* and *DEPEND*. Now I reverted to *DEPEND=virtual/jdk* and *RDEPEND=virtual/jre* and *pkg_setup* isn't needed anymore.

chewi commented on profiles/arch/arm/package.use.mask:
```diff
@@ -2,6 +2,10 @@
 
+# Thomas Deutschmann <whissi@gentoo.org> (2016-12-04)
+# sci-libs/gdal[mdb] requires dev-java/jackcess which is not keyworded
+>=sci-libs/gdal-2.0.3 mdb
```
I can test it on arm if you like. Not ppc though, we've dropped Java there.

Whissi replied:
For arm we have to keyword commons-lang and commons-logging first. So I will not keyword for jackcess for arm at the moment. Feel free to test & keyword all the dependencies once the package is in the repository.

chewi commented on dev-java/jackcess/jackcess-1.2.14.3.ebuild:
```diff
@@ -0,0 +1,44 @@
+
+LICENSE="Apache-2.0"
+SLOT="1"
+KEYWORDS="~amd64 ~x86 ~amd64-linux ~x86-linux ~ppc-macos ~x86-macos"
```
Ahem, where did all these come from!? :wink: You'll need ~ppc64 in here, I can test that if you like. As mentioned below, I can also test on arm.

Whissi replied:
When I started working on jackcess I think I copied these keywords from gdal ebuild and forget about it.

I added ~arm and ~ppc64.

Whissi commented:
All done, now. I am now waiting for another ping if this is ready or another change is requested.

chewi commented:
I've tested building jackcess on ppc64. I'm not going to test any more than that as you've dealt with all my feedback and I trust you know what you're doing. Go ahead!

Amynka commented:
To the older version of gdal 2 patches for jasper jpeg2k was added. It would be worth before commit to test if your new version bump works with new versions of jasper. 
https://gitweb.gentoo.org/repo/gentoo.git/tree/sci-libs/gdal/gdal-2.0.2-r3.ebuild#n88
I will try to test it in the mean time.

Whissi commented:
@Amynka: In gdal-2.0.3 we moved away from jasper (like Debian & Co. already did), see
```
 - Switched JPEG-2000 provider from media-libs/jasper to
   media-libs/openjpeg:2 because JasPer is unmaintained and vulnerable.
```

Amynka commented:
That is good news :)

Whissi commented:
OK, just for the status: I am now waiting for feedback regarding

>  I will try to test it in the mean time.

If everything is fine for you I'll move this into the repository.

Amynka commented:
I did test it and I can confirm it compiles fine. So go on.
